src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -301,25 +356,50 @@ public void onMessage(ConsumerRecord<K, V> record, Acknowledgment acknowledgment
 			Message<?> message = null;
 			try {
 				message = toMessagingMessage(record, acknowledgment);
+				setAttributesIfNecessary(record, message);
 			}
 			catch (RuntimeException e) {
-				Exception exception = new ConversionException("Failed to convert to message for: " + record, e);
-				if (getErrorChannel() != null) {
-					getMessagingTemplate().send(getErrorChannel(), new ErrorMessage(exception));
-				}
+				RuntimeException exception = new ConversionException("Failed to convert to message for: " + record, e);
+				sendErrorMessageIfNecessary(null, exception);
 			}
 			if (message != null) {
-				sendMessage(message);
+				try {
+					sendMessage(message);
+				}
+				finally {
+					attributesHolder.remove();
```

Comment:
M-m-m. I think this isn't correct when we are withing `RetryInterceptor`.
Our `onMessage()` will be called as much times as retry attempts and with the same `message` and `KAFKA_RAW_DATA`.
I don't see reason to remove and recreate the `TL` variable on each retry.
Might be better to leave the logic in the `close()` and remove here only if we don't have `retryTempalte`